### ISSUE TITLE
Audio playback fails to start with screen recorder apps with no indication of failure

### ISSUE THREAD

**Issue by fzurita** [google/oboe#1135]
Android version(s): Android 10
Android device(s): Huawei P30 Pro, plus other devices
Oboe version: 1.5.0
App name used for testing: M64Plus FZ
 
**Short description**
Starting playback while the built in screen recorder app is enabled produces no audio output.

**Steps to reproduce**
Start playback with built in screen recording app enabled.

**Expected behavior**
Audio plays

**Actual behavior**
Audio does not play

**Device**

```
ro.product.brand = HUAWEI
ro.product.manufacturer = HUAWEI
ro.product.model = VOG-L29
ro.product.device = HWVOG
ro.product.cpu.abi = arm64-v8a
ro.build.description = VOG-L29-user 10.0.0 HUAWEIVOG-L29 195-OVS-LGRP2 release-keys
ro.hardware = kirin980
ro.hardware.chipname = 
ro.arch = 
aaudio = 
[aaudio.mmap_exclusive_policy]: [2]
[aaudio.mmap_policy]: [2]

```

**Any additional context**
Even though there is no audio output and the logcat shows failures,```openManagedStream``` returns ```oboe::Result::OK``` and ```requestStart()``` also returns ```oboe::Result::OK```.

I have tried setting the stream to "shared" using the stream builder and it still silently fails. Changing the audio API to SLES allows audio to work.

As a workaround, I have a check where if ```getState() != oboe::StreamState::Started``` for a certain amount of time, then I switch the audio API to SLES.
[oboe_logcat_screen_recording.txt](https://github.com/google/oboe/files/5683327/oboe_logcat_screen_recording.txt)

The application is open source. Source code for how Oboe library interactions are performed is found here: https://github.com/mupen64plus-ae/mupen64plus-ae/blob/master/mupen64plus-audio-android/src/AudioHandler.cpp

**Comment by philburk** [comment]
@fzurita - Thanks for reporting this.

I see that this code is for an N64 emulator. http://www.mupen64plus.org/

> 	if (builder.openManagedStream(mOutStream) == oboe::Result::OK) {

We now recommend using std:shared_ptr<AudioStream> instead of ManagedStream but that is unrelated to this problem.

This may be related to [AAudioStreamBuilder_setAllowedCapturePolicy()](https://developer.android.com/ndk/reference/group/audio#aaudiostreambuilder_setallowedcapturepolicy).

It also might be related to MMAP. The Huawei P30 supports MMAP. But if you are using a sample rate that is not 48000 then you may not get MMAP.

> Starting playback while the built in screen recorder app is enabled produces no audio output.

I assume that means the game produces no audio. Is any audio recorded by the recorder?

I tried to reproduce this on R on Pixel 3a. I enabled screen recording using [these instructions](
https://support.google.com/pixelphone/answer/2811098?hl=en).
I recorded OboeTester with and without MMAP and it worked OK.

I then tested it on Q using:     adb shell settings put global settings_screenrecord_long_press true
and then long pressing Power button and then long pressing "ScreenShot".
OboeTester also worked OK.

Can you capture a short logcat from the open() and requestStart() calls? Please attach as a file, not pasted as text.

Also can you please try the TEST OUTPUT test from OboeTester? Then you can turn MMAP on/off. And try different sample rates.

**Comment by fzurita** [comment]
Hello @philburk The app is technically a front end to mupen64plus which can be found here: https://github.com/mupen64plus

The logcat output I have provided captures the events that you mentioned.

Also, I use this to get the sampling rate:
```AudioManager.getProperty(AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE)```

The above call is currently reporting 48000.

And yes, I should had been more specific. The game produces no audio, there is also no recorded audio. If I fallback to SLES, the game does output audio and audio is recorded as well.

I will try the app and report back.

**Comment by fzurita** [comment]
I compiled and ran the OboeTester. Same issue is apparent there. Setting 48000 as the frequency didn't help. I tried the most conservative settings possible and nothing worked until I changed the audio API to SLES. No errors are reported within the app unless I choose MMAP mode and only after I choose start. I will attach a logcat where conservative settings are used, but no audio is heard and no errors are reported. In the logcat, I did open, then start, then close while the screen recorded app is enabled.

[oboeTester_logcat.txt](https://github.com/google/oboe/files/5683777/oboeTester_logcat.txt)

**Comment by philburk** [comment]
Thanks for the logs. That really helps. I see:

2020-12-12 16:28:28.782 718-10284/? I/APM_AudioPolicyManager: sendAudioCloseOrOpen changed isAudioOpen=1
2020-12-12 16:28:28.783 29504-29504/com.google.sample.oboe.manualtest D/AAudio: AAudioStream_requestStart(s#12) returned 0 ---------

2020-12-12 16:28:28.784 29504-30005/com.google.sample.oboe.manualtest D/AudioStreamLegacy: onAudioDeviceUpdate() devId 3 => 0
2020-12-12 16:28:28.784 29504-30005/com.google.sample.oboe.manualtest D/AudioStreamLegacy: onAudioDeviceUpdate() request DISCONNECT in data callback due to device change

So I think the audio capture is triggering a routing change that is then causing a disconnect.
If you reconnect then you will get a stream but that will also get disconnected.

OpenSL ES handles device routing very differently so it does not have this problem.

This affects both Legacy and MMAP paths in AAudio in Q.

I think this is the same problem as #986

This is related to the internal bug:
b/139763500 | OboeTester plays for a half second then  disconnects if Live Caption is on
It was fixed in ag/11020894, which was merged into R.

In R and later, for AAudio, if you open a new stream immediately after the DISCONNECT then you will get a Legacy stream that will work OK.  We do not have a workaround for Q.

Closing because it is is fixed in R.

**Comment by fzurita** [comment]
Ok, thanks. I will continue with my workaround then.
